packages/gatsby-source-drupal/src/normalize.js

```diff
@@ -12,11 +13,132 @@ const getHref = link => {
 
 exports.getHref = getHref
 
-const nodeFromData = (datum, createNodeId, entityReferenceRevisions = []) => {
+const imageCDNState = {
+  foundPlaceholderStyle: false,
+  hasLoggedNoPlaceholderStyle: false,
+}
+
+exports.imageCDNState = imageCDNState
+
+/**
+ * This FN takes in node data and returns Gatsby Image CDN fields that should be added to that node. If the input node isn't an image an empty object is returned.
+ */
+const getGatsbyImageCdnFields = async ({
+  node,
+  type,
+  pluginOptions,
+  fileNodesExtendedData,
+  reporter,
+}) => {
+  if (!pluginOptions?.skipFileDownloads) {
+    return {}
+  }
+
+  const isFile = isFileNode({
+    internal: {
+      type,
+    },
+  })
+
+  if (!isFile) {
+    return {}
+  }
+
+  const mimeType = node.attributes.filemime
+
+  if (!mimeType) {
+    return {}
+  }
+
+  const url = getFileUrl(node.attributes, pluginOptions.baseUrl)?.href
+
+  if (!url) {
+    return {}
+  }
+
+  if (!mimeType.includes(`image/`)) {
+    return {
+      mimeType,
+      url,
+    }
+  }
+
+  const extraNodeData = fileNodesExtendedData?.get(node.id) || null
+
+  try {
+    const { placeholderStyleName } = getOptions()
+    const placeholderUrl =
+      extraNodeData?.imageDerivatives?.links?.[placeholderStyleName]?.href ||
+      extraNodeData?.imageDerivatives?.links?.placeholder?.href ||
+      url
+
+    if (placeholderUrl !== url) {
+      imageCDNState.foundPlaceholderStyle = true
+    }
+
+    const hasRequiredData = input => input && input.width && input.height
+
+    const imageSize = hasRequiredData(extraNodeData)
+      ? extraNodeData
+      : await probeImageSize(url)
```

Comment:
Can you maybe add a comment when axtraNodeData is available and when probeImageSize is being used? Just for my reference. Is there a way to get rid of probeImageSize?